Sources/Smithy/Context.swift

```diff
@@ -5,46 +5,63 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+import class Foundation.NSRecursiveLock
+
 public class Context {
-    public var attributes: Attributes
+    private var _attributes: Attributes
+    private let lock = NSRecursiveLock()
 
     public init(attributes: Attributes) {
-        self.attributes = attributes
+        self._attributes = attributes
     }
 
     public func toBuilder() -> ContextBuilder {
-        let builder = ContextBuilder()
-        builder.attributes = self.attributes
-        return builder
+        ContextBuilder(attributes: accessAttributes())
     }
 
     public func getLogger() -> LogAgent? {
-        return attributes.get(key: AttributeKeys.logger)
+        return accessAttributes().get(key: AttributeKeys.logger)
+    }
+
+    @discardableResult
+    private func accessAttributes(accessor: ((inout Attributes) -> Void)? = nil) -> Attributes {
+        lock.lock()
+        defer { lock.unlock() }
+        accessor?(&_attributes)
+        return _attributes
     }
 }
```

Comment:
The `accessAttributes` function above enforces exclusive access to the now-private `_attributes` property.